src/librustdoc/clean/mod.rs

```diff
@@ -1698,11 +1717,197 @@ fn normalize<'tcx>(
+            // we are in a non-expression one, it should be `'static`
+            // This check is only correct if we are *not* in an *expression context*!
+            // If we are in one though, then only return `None` if the lifetime equals `'_`.
+            if lifetime == Lifetime::statik() { None } else { Some(lifetime) }
+        }
+        // > If the trait is defined with a single lifetime bound then that bound is used.
+        [trait_region] => match *trait_region {
+            // > If 'static is used for any lifetime bound then 'static is used.
+            ty::ReStatic if lifetime == Lifetime::statik() => None,
+            ty::ReEarlyBound(region) if lifetime.0 == region.name => None,
+            // FIXME(fmease): The following branch is obviously incorrect. Adding the guard leads to
+            // `'static` not being elided in `dyn for<'any> Bounded0<'any>`. Not sure yet what the
+            // actual condition has to be.
+            #[allow(unused_variables)]
+            ty::ReLateBound(_, ty::BoundRegion { kind: ty::BrNamed(_, region), .. })
+                /*if lifetime.0 == region*/ =>
+            {
+                None
+            }
+            _ => return Some(lifetime),
+        },
+        // There are several distinct trait regions and none are `'static` (thanks to the preprocessing above).
+        // Due to ambiguity there is no default trait-object lifetime and thus elision is impossible.
+        // Don't elide the lifetime.
+        _ => Some(lifetime),
+    }
+}
+
+#[derive(Debug)]
+pub(crate) enum ContainerTy<'tcx> {
+    Ref(ty::Region<'tcx>),
+    Regular { ty: DefId, substs: ty::Binder<'tcx, &'tcx [ty::GenericArg<'tcx>]>, arg: usize },
+}
+
+impl<'tcx> ContainerTy<'tcx> {
+    fn object_lifetime_default(self, tcx: TyCtxt<'tcx>) -> ObjectLifetimeDefault<'tcx> {
+        match self {
+            Self::Ref(region) => ObjectLifetimeDefault::Arg(region),
+            Self::Regular { ty: container, substs, arg: index } => {
+                // FIXME(fmease): Which kinds do we actually want to allow?
+                let (DefKind::TyAlias
+                | DefKind::Struct
+                | DefKind::Union
+                | DefKind::Enum
+                | DefKind::ForeignTy
+                | DefKind::AssocTy) = tcx.def_kind(container)
+                else {
+                    return ObjectLifetimeDefault::Empty;
+                };
+
+                let generics = tcx.generics_of(container);
+                let param = generics.params[index].def_id;
+                let default = tcx.object_lifetime_default(param);
+
+                match default {
+                    rbv::ObjectLifetimeDefault::Param(lifetime) => {
+                        // FIXME(fmease): Don't use `skip_binder` here! Also this indexing dance looks
+                        // quite fragile and I'm probably misusing the API.
+                        let arg = substs.skip_binder()
```

Comment:
This indexing dance is ok, that's what `rustc_hir_analysis::collect::resolve_bound_vars::visit_segment_args` does.